### ISSUE TITLE
build error reporting libbinlog.a file format not recognized

### ISSUE THREAD

**Issue by welaunchcn** [morganstanley/binlog#136]
I followed the install steps but the following error is reporeted:
[ 21%] Linking CXX executable brecovery
/usr/bin/cmake -E cmake_link_script CMakeFiles/brecovery.dir/link.txt --verbose=1
/usr/bin/c++  -O2 -g -DNDEBUG   CMakeFiles/brecovery.dir/bin/brecovery.cpp.o  -o brecovery  libbinlog.a 
/usr/bin/ld: libbinlog.a: error adding symbols: file format not recognized
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CMakeFiles/brecovery.dir/build.make:85: brecovery] Error 1
make[2]: Leaving directory '/home/user/project/public/binlog/Release'
make[1]: *** [CMakeFiles/Makefile2:571: CMakeFiles/brecovery.dir/all] Error 2
make[1]: Leaving directory '/home/user/project/public/binlog/Release'
make: *** [Makefile:141: all] Error 2

libbinlog.a has been generated

**Comment by erenon** [comment]
Please post the output of `/usr/bin/c++ --version` and `/usr/bin/ld --version`. Perhaps these two are for different architectures? Can you compile other projects with `/usr/bin/c++` ?

**Comment by erenon** [comment]
From the `make VERBOSE=1` output, please show how `libbinlog.a` was built.

**Comment by welaunchcn** [comment]
Please find the full cmake and build log attached.
[buildlog.txt](https://github.com/morganstanley/binlog/files/7547942/buildlog.txt)

**Comment by erenon** [comment]
It appears to be an incompatibility between llvm-ar + -fthin-lto and the gnu linker.
To verify that LTO is at fault, in CMakeLists.txt, please remove this block:

```
# Interprocedural Optimization (LTO)
if(POLICY CMP0069)
  cmake_policy(SET CMP0069 NEW)
  include(CheckIPOSupported) # fails if policy 69 is not set
  check_ipo_supported(RESULT BINLOG_HAS_IPO)
  if(BINLOG_HAS_IPO)
    message(STATUS "Use interprocedural optimization")
  endif()
endif()
```

And instead add:

```
set(BINLOG_HAS_IPO OFF)
```

If this makes it build, then the proper solution is either:

 - force clang to use llvm linkers (`-fuse-ld=lld`, perhaps)
 - force clang to use gnu ar exclusively (don't know how)

**Comment by welaunchcn** [comment]
Yes, removing that block does make it work. Thank you very much for the help.

**Comment by erenon** [comment]
Thanks for confirming. I renamed the ticket to reflect the uncovered issue. The LTO checker should detect this degenerate case, and disable LTO automatically.
